// component-router/lib/actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.removeRoute = exports.addRoute = exports.removeParam = exports.addDefaultParam = exports.restoreLocation = exports.navigateTo = undefined;

var _Constants = require('./Constants');

var _Constants2 = _interopRequireDefault(_Constants);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var navigateTo = exports.navigateTo = function navigateTo(_ref) {
  var pathname = _ref.pathname;
  var query = _ref.query;
  return {
    type: _Constants2.default.NAVIGATE_TO,
    payload: { pathname: pathname, query: query }
  };
};

var restoreLocation = exports.restoreLocation = function restoreLocation(_ref2, locationType) {
  var pathname = _ref2.pathname;
  var search = _ref2.search;
  var hash = _ref2.hash;
  return {
    type: _Constants2.default.RESTORE_LOCATION,
    payload: { location: { pathname: pathname, search: search, hash: hash }, locationType: locationType }
  };
};

var addDefaultParam = exports.addDefaultParam = function addDefaultParam(namespace, value) {
  return {
    type: _Constants2.default.ADD_DEFAULT_PARAM,
    payload: { namespace: namespace, value: value }
  };
};

var removeParam = exports.removeParam = function removeParam(namespace) {
  return {
    type: _Constants2.default.REMOVE_PARAM,
    payload: { namespace: namespace }
  };
};

var addRoute = exports.addRoute = function addRoute(route) {
  return {
    type: _Constants2.default.ADD_ROUTE,
    payload: { route: route }
  };
};

var removeRoute = exports.removeRoute = function removeRoute(route) {
  return {
    type: _Constants2.default.REMOVE_ROUTE,
    payload: { route: route }
  };
};
