// component-router/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.actions = exports.Constants = exports.locationNode = exports.locationHistory = exports.locationHash = exports.location = exports.isActive = exports.href = exports.componentRouter = undefined;

var _reducer = require('./reducer');

Object.defineProperty(exports, 'componentRouter', {
  enumerable: true,
  get: function get() {
    return _reducer.componentRouter;
  }
});
Object.defineProperty(exports, 'href', {
  enumerable: true,
  get: function get() {
    return _reducer.href;
  }
});
Object.defineProperty(exports, 'isActive', {
  enumerable: true,
  get: function get() {
    return _reducer.isActive;
  }
});

var _location = require('./location');

Object.defineProperty(exports, 'location', {
  enumerable: true,
  get: function get() {
    return _location.location;
  }
});

var _hash = require('./adapters/hash');

Object.defineProperty(exports, 'locationHash', {
  enumerable: true,
  get: function get() {
    return _hash.locationHash;
  }
});

var _history = require('./adapters/history');

Object.defineProperty(exports, 'locationHistory', {
  enumerable: true,
  get: function get() {
    return _history.locationHistory;
  }
});

var _node = require('./adapters/node');

Object.defineProperty(exports, 'locationNode', {
  enumerable: true,
  get: function get() {
    return _node.locationNode;
  }
});

var _Constants = require('./Constants');

Object.defineProperty(exports, 'Constants', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_Constants).default;
  }
});

var _actions = require('./actions');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var actions = exports.actions = {
  navigateTo: _actions.navigateTo, restoreLocation: _actions.restoreLocation, addDefaultParam: _actions.addDefaultParam, removeParam: _actions.removeParam, addRoute: _actions.addRoute, removeRoute: _actions.removeRoute
};
